### ISSUE TITLE
[regression] testing fails with 0.3

### ISSUE THREAD

**Issue by zetok** [BurntSushi/quickcheck#141]
Error:
```
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Error { repr: Os { code: 11, message: "Resource temporarily unavailable" } }', ../src/libcore/result.rs:746
```

full output: https://gist.github.com/zetok/b8a1d8981927c67ad789092b046d1c33

Not only tests error out, but also tests don't want to finish – usually it takes ~2 min for tests to run & pass, while after bumping version from `0.2` to `0.3` tests didn't finish even after 15 min, and the process seemed to be idle.

Tested under Rust 1.9
lib: https://github.com/zetok/tox ← just bump version to 0.3 to reproduce
probably the offending commit: 604a0cea637b3582425474de847d8e3fdc78d96e

**Comment by BurntSushi** [comment]
Hmm, unfortunately, I can't reproduce this issue on Rust nightly (1.13, ef9786ce0, 2016-09-02) and tests run in about the same time regardless of quickcheck `0.2` or `0.3`.

I tried the same steps on Rust 1.9 and I can't reproduce your issue either. Again, tests run in about the same time. (25 seconds.)

What OS are you on? (I'm on Linux.)

**Comment by zetok** [comment]
Linux.

**Comment by BurntSushi** [comment]
Also, I have no clue what specifically would be causing IO errors, so I'm out of ideas on this one without more data. (Although I'm not sure what "more data" means.)

**Comment by BurntSushi** [comment]
You might try running with `RUST_BACKTRACE=1` to see where those unwraps are triggering.

**Comment by zetok** [comment]
with `RUST_BACKTRACE=`: https://gist.github.com/zetok/832a3925bd34f861df88b8279974eff5

Not sure if it matters, but I'm running Hardened Gentoo with whole OS compiled with `-O2 -march=native -ggdb -pipe -fstack-protector-strong`.

Aside from that, I also had 
```
# cat /proc/sys/kernel/pid_max 
9999
```

I've changed it "back" to 32k with
```
# sysctl kernel.pid_max=32000
kernel.pid_max = 32000
```
but the result was the same.

**Comment by BurntSushi** [comment]
Looking at a backtrace chosen at random:

```
stack backtrace:
   1:     0x564f3a09c880 - std::sys::backtrace::tracing::imp::write::h4c73fcd3363076f5
   2:     0x564f3a09fd9b - std::panicking::default_hook::_$u7b$$u7b$closure$u7d$$u7d$::h0422dbb3077e6747
   3:     0x564f3a09fa23 - std::panicking::default_hook::haac48fa641db8fa2
   4:     0x564f3a09087f - std::sys_common::unwind::begin_unwind_inner::h39d40f52add53ef7
   5:     0x564f3a091868 - std::sys_common::unwind::begin_unwind_fmt::h64c0ff793199cc1b
   6:     0x564f3a09bad1 - rust_begin_unwind
   7:     0x564f3a0d620f - core::panicking::panic_fmt::h73bf9d7e8e891a73
   8:     0x564f39cd6b47 - core::result::unwrap_failed::h8b8227d746f1a946
                        at ../src/libcore/macros.rs:29
   9:     0x564f39cd5e2f - _<std..result..Result<T, E>>::unwrap::h3d1955a5e90e8a20
                        at ../src/libcore/result.rs:687
  10:     0x564f39cd84a8 - std::thread::spawn::h8b45c215e4e5e8a8
                        at ../src/libstd/thread/mod.rs:320
  11:     0x564f39cd7f64 - tox::toxcore_tests::crypto_core_tests::crypto_init_test::with_threads::_$u7b$$u7b$closure$u7d$$u7d$::h23004a1aecd3bfaa
                        at src/toxcore_tests/crypto_core_tests.rs:44
  12:     0x564f39cd7c8f - std::thread::Builder::spawn::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::he38dfee3ca763109
  13:     0x564f39cd7c69 - std::sys_common::unwind::try::try_fn::hfd7b2168cc6a67fa
                        at ../src/libstd/sys/common/unwind/mod.rs:127
  14:     0x564f3a09ba5b - __rust_try
  15:     0x564f3a09b9ed - std::sys_common::unwind::inner_try::h9eebd8dc83f388a6
  16:     0x564f39cd7c20 - std::sys_common::unwind::try::h095fa1016dff63c1
                        at ../src/libstd/sys/common/unwind/mod.rs:123
  17:     0x564f39cd7ad1 - std::thread::Builder::spawn::_$u7b$$u7b$closure$u7d$$u7d$::ha16faf11cddfb033
  18:     0x564f39cd8178 - _<F as std..boxed..FnBox<A>>::call_box::hfb0cfe38e43ac8cd
                        at ../src/liballoc/boxed.rs:541
  19:     0x564f3a09e444 - std::sys::thread::Thread::new::thread_start::h471ad90789353b5b
  20:     0x7f79c9a76613 - <unknown>
  21:     0x7f79c95a1f6c - __clone
  22:                0x0 - <unknown>
```

It looks like `tox::toxcore_tests::crypto_core_tests::crypto_init_test::with_threads` is trying to spawn a thread and it's failing. I'm not quite sure why quickcheck `0.3` would provoke this over quickcheck `0.2`, but I'm not convinced there's a regression here. (For context, `0.2` would spawn a new thread for every test that it ran while `0.3` runs every test in the same thread and catches panics using the new unwind API. If anything, I'd imagine `0.3` would help the issue, not agitate it.)

**Comment by zetok** [comment]
Oh, right, I also have set `ulimit -u 5000`.
`with_threads()` can on its own spawn up to 1020 threads. Lowering to 255 "fixes" it for me with `0.3`.

**Comment by BurntSushi** [comment]
Thanks for the update! Seems like all is well then.
